[1.14.x] Cherry-pick network fix from 1.15

pau101 opened the pull request:
Backports https://github.com/MinecraftForge/MinecraftForge/commit/aca45340bfd1ae8de9857cb906a2a31fbf8fbcdb to 1.14

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=10742) <br/>All committers have signed the CLA.

LexManos commented:
You beed to sign the cla.
Also its not a simple cherry pick.
You have to maintain binary compatibility. Which means no removing methods. So you need to add a bouncer with sane defaults if you add a new argument to a method.

Jonathing reviewed (changes requested):
Same as [Lex's comment](https://github.com/MinecraftForge/MinecraftForge/pull/10742#issuecomment-3698682331).

pau101 committed:
Fix up alarming security crisis with network handling that allowed for wrong code execution on the server, resulting in CRASHED SERVERS. AWFUL stuff.

Also fixed a silly log message from the login handler.

This change introduces a mechanism to direct certain packets to only process on one side or another. Invalid sidedness will result in the connection being terminated.

pau101 commented:
Binary compatibility is maintained, unless reflection into private members is included in what counts as relevant? The `MessageHandler` is a package-private inner class in `IndexedMessageCodec` if that is what were you seeing. As well `addCodecIndex` is also a package-private internal method.

The CLA was signed, had to commit without cherry-pick.

Jonathing commented:
Yes, reflection into private methods count, especially for versions older than the latest (currently 1.21.11). For older versions, we assume the worst in terms of what mods will do with coremodding and hacks.

pau101 committed:
Readd network preexisting methods

pau101 commented:
Changes made!

Spectrall368 commented:
From what I've read about the backport this might fix https://github.com/MinecraftForge/MinecraftForge/issues/10588

<img width="1029" height="231" alt="image" src="https://github.com/user-attachments/assets/f4eb3edc-0565-4085-b3ce-302fed289678" />

pau101 commented:
No that other issue is a client crash, not server, and unrelated.

Spectrall368 commented:
Ok

LexManos commented:
> Binary compatibility is maintained, unless reflection into private members is included in what counts as relevant? 

It is, welcome to the hell that is old Minecraft modding. Modders hack into internals all the time and I'd rather not break them in such old versions. We even have to sometimes do hacks such as keeping dead code alive because people edit forge's bytecode during loading.

LexManos commented:
Alright, so your code doesn't work/compile. So guess i'm just gunna have to do this myself.

LexManos commented:
There, done in a purely additive way, only code change was implementation of onCustomPayload and that's because the optional BS bugged me.

pau101 commented:
Thanks! Nice to know this is resolved.
